Remove 'tensor' of a joined TensorMap if possible

PicoCentauri opened the pull request:
Adds the functionality to `join` to remove the `"tensor"` column name while joining TensorMaps.

Due to torch-script I can not use a try-catch block like

```python
try:
    return remove_labels_column(tensor_joined, name="tensor", axis=axis)
except EquistoreError:
    return tensor_joined
```

Instead, I have to check the uniqueness on the Python level before calling `remove_labels_column`. If someone has an idea to get around the his I am happy to change this.

jwa7 commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -126,7 +160,22 @@ def join(tensors: List[TensorMap], axis: str):
+    # Remove "tensor" Labels dimension if a new Labels of all blocks would be unique
+    # Due to toch script we can use a try-except block but have to check in advance if
+    # the `remove_dimension` would be succesful, i.e the new Labels instance is unique.
+    for block in tensor_joined:
+        if axis == "samples":
+            n_values = len(block.samples)
+            # "tensor" names is inserted at the end of the Labels dimensions
+            new_names = np.delete(block.samples, -1, axis=1)
+        elif axis == "properties":
+            n_values = len(block.properties)
+            # "tensor" name is inserted at the begininng of the Labels dimensions
+            new_names = np.delete(block.properties, 0, axis=1)
+
+        n_values_unique = len(np.unique(new_names, axis=0))
+
+        if n_values != n_values_unique:
+            return tensor_joined
```
Perhaps `unique_metadata` can be used here?

```suggestion
    if len(equistore.unique_metadata(tensor_joined, axis=axis, name="tensor")) > 1:
        return tensor_joined
```

jwa7 replied:
Hmm apologies maybe not: see https://github.com/lab-cosmo/equistore/pull/314#discussion_r1273274348

PicoCentauri replied:
Yes as we talked yesterday I will try to use Labels::intersection.

agoscinski commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
+                        [0, 2, 3, 1, 4, 5]
+
+    2. If Labels names are the same but the values are not unique, a new dimension
+       ``"tensor"`` is added to the names::
```
```suggestion
       ``"tensor"`` is added at the beginning to the names which determines from which TensorMap the label entry originally comes from ::
```

maybe a comment on where the new name is added, always at the beginning?

agoscinski commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
+    1. If Labels names are the same and the values are unique, we keep the names and
+       join the values::
+
+        labels_1 = ["n"], [0, 2, 3]
```
This could be actually super nice doctests. not enforcing this, but would be nice

PicoCentauri replied:
Yes, I though the same. I think I can do it!

agoscinski commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
+    of the three following strategies:
+
+    1. If Labels names are the same and the values are unique, we keep the names and
+       join the values::
```
why two double colons at the end? is this related to sphinx?

PicoCentauri replied:
Yes this renders the next paragraph as code.

agoscinski commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
+            joined_labels = ["tensor", "n"],
+                            [[0, 0], [0, 2], [0, 3], [1, 0], [1, 2]]
+
+    3. If Labels names are different we change the names to ("tensor", "property"). This
```
```suggestion
    3. If Labels names are different we change the names to ("tensor", "property"). The column "property" corresponds to the ith label entry TensorMap specified by tensor. This
```

agoscinski commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -126,7 +160,22 @@ def join(tensors: List[TensorMap], axis: str):
-
-    return tensor_joined
+    # Remove "tensor" Labels dimension if a new Labels of all blocks would be unique
+    # Due to toch script we can use a try-except block but have to check in advance if
```
```suggestion
    # Due to toch script we cant use a try-except block but have to check in advance if
```

agoscinski reviewed (commented):
If we input a TensorMap with a label column name "tensor" or "property" can we raise an error already in this function? Something like 
```
raise ValueError(f"join cannot be applied on TensorMaps with a column name 'tensor' in the TensorMap keys or in targeted axis {axis}. Please rename the corresponding column first using ....")
```

agoscinski commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
+
+       will lead to::
+
+            joined_labels = ["tensor", "property"],
```
What do you think about using a notation for "private" variables in python for metadata column names that can exclusively only used equistore for operations "_eqs_tensor" and "_eqs_property". To be honest, I don't have any other ideas of applications than this function, but it could be already here useful.

PicoCentauri replied:
Hmm I think private names will make everything harder to follow.

agoscinski replied:
They actually do this in pandas for their merge function
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html
> **indicator: bool or str, default False**
>
> If True, **adds a column to the output DataFrame called “_merge” with information on the source of each row**. The column can be given a different name by providing a string argument. The column will have a Categorical type with the value of “left_only” for observations whose merge key only appears in the left DataFrame, “right_only” for observations whose merge key only appears in the right DataFrame, and “both” if the observation’s merge key is found in both DataFrames.

agoscinski commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
     `properties` dimension and for ``axis='samples'`` they will be the along the
```
```
`axis='properties'`` it will be the `tensor_maps` will be joined along the
```
->
```
`axis='properties'`` the tensor maps in `tensors` will be joined along the
```

PicoCentauri commented on python/equistore-operations/equistore/operations/join.py:
```diff
@@ -7,6 +7,7 @@
+from .manipulate_dimension import remove_dimension
 
 
 def join(tensors: List[TensorMap], axis: str):
```
Yes it is more `concatenate` but I would say we keep it like this and can rename the bunch when we have more operations.

PicoCentauri commented:
I added an new boolean parameter `remove_tensor_name` to control if the tensor dimension should be removed or not. We can also leave this out and completly and only put a comment that the user can remove the tensor using the `equistore.remove_dimension`. To be honest I am more and more in favor to do this. What do the reviewers think @jwa7, @agoscinski  and @Luthaf 

Also I added your the examples you want @agoscinski. Hope they are fine and understandable. I tried to keep the code to create the tensors as small as possible.

jwa7 reviewed (commented):
I like the having the bool option to remove, but think we should have a default to False. Thoughts?

agoscinski commented:
> I added an new boolean parameter remove_tensor_name to control if the tensor dimension should be removed or not. We can also leave this out and completly and only put a comment that the user can remove the tensor using the equistore.remove_dimension. To be honest I am more and more in favor to do this. What do the reviewers think @jwa7, @agoscinski and @Luthaf

Given the discussion in https://github.com/lab-cosmo/equistore/issues/315 we will break the API of join anyway, I dont think it matters much what we choose now. Let's just go with an option and set it to False as @jwa7 suggests it.
